server/src/main/java/org/elasticsearch/index/mapper/ContentPath.java

```diff
@@ -8,55 +8,47 @@
 
 package org.elasticsearch.index.mapper;
 
+import org.elasticsearch.common.Strings;
+
 public final class ContentPath {
 
     private static final char DELIMITER = '.';
 
-    private final StringBuilder sb;
-
-    private final int offset;
+    private static final ThreadLocal<StringBuilder> stringBuilder = ThreadLocal.withInitial(StringBuilder::new);
 
-    private int index = 0;
+    private int index;
 
-    private String[] path = new String[10];
+    private String[] path = Strings.EMPTY_ARRAY;
 
     public ContentPath() {
-        this(0);
-    }
-
-    /**
-     * Constructs a json path with an offset. The offset will result an {@code offset}
-     * number of path elements to not be included in {@link #pathAsText(String)}.
-     */
-    public ContentPath(int offset) {
-        this.sb = new StringBuilder();
-        this.offset = offset;
         this.index = 0;
     }
 
     public ContentPath(String path) {
-        this.sb = new StringBuilder();
-        this.offset = 0;
         this.index = 0;
         add(path);
     }
 
     public void add(String name) {
-        path[index++] = name;
         if (index == path.length) { // expand if needed
             String[] newPath = new String[path.length + 10];
             System.arraycopy(path, 0, newPath, 0, path.length);
             path = newPath;
         }
+        path[index++] = name;
     }
 
     public void remove() {
         path[index--] = null;
     }
 
     public String pathAsText(String name) {
+        if (index == 0) {
+            return name;
+        }
```

Comment:
Can we try it with just this shortcut? Threadlocals scare me and I reckon that this change alone will help with the vast majority of cases.